### PR TITLE
Remove tree validations and introduce ParameterizedQuery

### PR DESCRIPTION
**Pull request by rauchy**
This PR splits removes `SQLQuery` and all tree-validation-related logic and introduces `ParameterizedQuery`, which handles parameter validation and templating.

### CODE REVIEW

File: redash/handlers/query_results.py [hunk 1]
```diff
@@ -10,90 +10,59 @@
...
+    parameterized_query = parameterized_query_class(query_text).apply(parameter_values)
 
-    if max_age <= 0:
-        query_result = None
-    else:
-        query_result = models.QueryResult.get_latest(data_source, query_text, max_age)
+    if parameterized_query.missing_params:
+        raise Exception('Missing parameter value for: {}'.format(", ".join(parameterized_query.missing_params)))
 
+    query_text = parameterized_query.query
```
Comment (rauchy) [inline redash/handlers/query_results.py]: Note that I'm not calling `parameterized_query.text` here, hence there's no attempt to detect SQL injection for this case.

File: redash/query_runner/google_spreadsheets.py [hunk 2]
```diff
@@ -147,6 +147,9 @@ def request(self, *args, **kwargs):
 
 
 class GoogleSpreadsheet(BaseQueryRunner):
+    def __init__(self, configuration):
+        super(GoogleSpreadsheet, self).__init__(configuration)
+        self.syntax = 'custom'
```
Comment (arikfr) [inline redash/query_runner/google_spreadsheets.py]: Did you check that the UI doesn't blow with syntax names it doesn't recognize?

Comment (rauchy) [reply redash/query_runner/google_spreadsheets.py]: Seems like this only affects query formatting. I tested it by changing `BaseQueryRunner`'s syntax to `custom` and noticed no explosions other than query formatting.

File: redash/handlers/query_results.py [hunk 3]
```diff
@@ -10,90 +10,59 @@
...
 def run_query_sync(data_source, parameter_values, query_text, max_age=0):
-    missing_params = find_missing_params(query_text, parameter_values)
-    if missing_params:
-        raise Exception('Missing parameter value for: {}'.format(", ".join(missing_params)))
-
-    query_text = apply_parameters(query_text, parameter_values, data_source)
+    parameterized_query_class = data_source.query_runner.parameterized_query_class
+    parameterized_query = parameterized_query_class(query_text).apply(parameter_values)
 
-    if max_age <= 0:
```
[the remark this answers is not in the document]
Comment (rauchy) [reply redash/handlers/query_results.py]: Which change in logic are you referring to?

File: redash/handlers/query_results.py [hunk 4]
```diff
@@ -102,17 +71,37 @@ def run_query(data_source, parameter_values, query_text, query_id, max_age=0):
...
+            'parameters': parameter_values,
+            'timestamp': time.time(),
+            'org_id': data_source.org_id
+        })
+        query_text = parameterized_query.query
+    except Exception as e:
+        logging.info(u"Failed applying parameters for query %s: %s", gen_query_hash(parameterized_query.query), e.message)
+        query_text = parameterized_query.query
+
+    if max_age:
```
[the remark this answers is not in the document]
Comment (rauchy) [reply redash/handlers/query_results.py]: Ahh... ok, reverting.

**Pull request merged**